staging/src/k8s.io/client-go/tools/cache/reflector_data_consistency_detector.go

```diff
@@ -32,42 +33,46 @@ import (
 	"k8s.io/klog/v2"
 )
 
-var dataConsistencyDetectionEnabled = false
+var dataConsistencyDetectionForWatchListEnabled = false
 
 func init() {
-	dataConsistencyDetectionEnabled, _ = strconv.ParseBool(os.Getenv("KUBE_WATCHLIST_INCONSISTENCY_DETECTOR"))
+	dataConsistencyDetectionForWatchListEnabled, _ = strconv.ParseBool(os.Getenv("KUBE_WATCHLIST_INCONSISTENCY_DETECTOR"))
 }
 
-// checkWatchListConsistencyIfRequested performs a data consistency check only when
+type retrieveCollectedItemsFunc[U any] func() []U
+
+type listFunc[T runtime.Object] func(ctx context.Context, options metav1.ListOptions) (T, error)
+
+// checkWatchListDataConsistencyIfRequested performs a data consistency check only when
 // the KUBE_WATCHLIST_INCONSISTENCY_DETECTOR environment variable was set during a binary startup.
 //
 // The consistency check is meant to be enforced only in the CI, not in production.
 // The check ensures that data retrieved by the watch-list api call
-// is exactly the same as data received by the standard list api call.
+// is exactly the same as data received by the standard list api call against etcd.
 //
 // Note that this function will panic when data inconsistency is detected.
 // This is intentional because we want to catch it in the CI.
-func checkWatchListConsistencyIfRequested(stopCh <-chan struct{}, identity string, lastSyncedResourceVersion string, listerWatcher Lister, store Store) {
-	if !dataConsistencyDetectionEnabled {
+func checkWatchListDataConsistencyIfRequested[T runtime.Object, U any](ctx context.Context, identity string, lastSyncedResourceVersion string, listFn listFunc[T], retrieveCollectedItemsFn retrieveCollectedItemsFunc[U]) {
+	if !dataConsistencyDetectionForWatchListEnabled {
 		return
 	}
-	checkWatchListConsistency(stopCh, identity, lastSyncedResourceVersion, listerWatcher, store)
+	// for informers we pass an empty ListOptions because
```

Comment:
not sure I follow this comment

but generally, I would like to use the exact same checker for ConsistentReadsFromCache - compare the result returned from watchcache with the one that etcd would return

Reply:
I want the checking to be visibly simpler, and I would probably would like to do that server-side actually (probably directly in watchcache actually). But that's a follow-up.
At high-level, we're on the same page here.

Reply:
> I would probably would like to do that server-side actually

Actually, I discussed adding the check to the watch-cache/server-side with Marek. We figured it would be better to have it on the client-side since we already have that for informers/reflectors and we wanted to reuse the same code. Also by having it on the client-side we also kind of testing our APIs.